File: pulso/apify.py
"""Actores de Apify: raspado de plataformas sociales sin iniciar sesion.

Apify es una plataforma de actores rentados: se le paga por corrida y ella
opera el navegador. Eso NO cambia el analisis legal de docs/PLAN.md seccion 3,
y conviene decirlo aqui porque es la confusion natural: delegar el navegador
no delega la responsabilidad. En Meta v. Bright Data (N.D. Cal., enero 2024) la
defensa que prospero dependio de NO ser 'usuario' de la plataforma, o sea de
no haber iniciado sesion. Un actor que se loguea destruye esa defensa igual de
bien si el login lo hace un tercero por contrato, y encima agrega la cuenta
del proveedor al expediente. Meta v. Voyager Labs es ese patron de hechos.

De ahi las tres reglas que este modulo vuelve ejecutables en vez de dejarlas
como comentario:

1. **Ningun actor con sesion.** `revisar_entrada()` rechaza una entrada que
   traiga cookies, credenciales o tokens de sesion (LLAVES_DE_SESION). Es un
   error de configuracion, no un aviso: un actor logueado no se degrada, se
   prohibe. La lista deja fuera `username` y `user` a proposito, porque en
   casi todos los actores son el perfil OBJETIVO y no una credencial.

2. **El texto crudo no entra a git.** Misma disciplina que pulso/conversacion.py,
   pero mas estricta: YouTube al menos concede 30 dias por politica escrita
   (III.E.4.d) y Meta, TikTok y X no conceden nada. Lo que ata aqui son los
   terminos de cada plataforma mas la LFPDPPP mexicana y la CPRA californiana,
   porque un comentario con nombre propio es dato personal en las dos. El
   crudo vive en cache/ (ignorado por git, TTL de 30 dias) y a data/ solo
   llegan conteos derivados.

3. **El presupuesto es un tope duro, no una sugerencia.** Apify cobra por
   resultado y por unidad de computo. El cron corre cuatro veces al dia: un
   `maxItems` sin vigilar no produce una corrida lenta, produce una factura.
   El presupuesto se REPARTE entre los actores activos en vez de gastarse en
   orden de archivo -- es la misma leccion que pulso/busquedas.py aprendio
   cuando las dos primeras consultas se llevaron las 40 peticiones y las otras
   cuatro quedaron en cero.

## El tope de 300 segundos

`run-sync-get-dataset-items` corta a los 300 segundos con HTTP 408. No se
sube: es limite de la plataforma. Un raspado grande de Instagram lo pasa sin
esfuerzo, asi que el tope de resultados por actor esta puesto para caber
adentro. Si un actor empieza a devolver 408, la respuesta correcta es bajar su
cuota en config/apify.json, no subir el timeout ni pasar al modo asincrono
callado: una corrida que tarda mas de lo que el cron tolera es una corrida que
gasta dinero y no commitea nada.

## Idioma, y por que cada actor lo declara

El modelo de tono es espanol (RoBERTuito, tweets de TASS). El lado San Diego
publica en ingles. Un modelo de sentimiento al que se le da texto en otro
idioma no devuelve un error, devuelve una etiqueta plausible, y asi estuvo
roto meses del lado de la prensa. Cada actor declara `idioma` en la
configuracion y nunca se adivina del texto.

## Zona

Un actor no lleva `zona`, por la misma razon que una busqueda no la lleva
(ver pulso/busquedas.py): una cuenta o un hashtag le acreditarian esa zona a
todo resultado que no nombre lugar alguno, que es el bug de El Imparcial y
Hermosillo con otro disfraz. Las fuentes sinteticas son `estatal`.

## Degradacion

Sin `APIFY_TOKEN` el paso no truena: devuelve vacio y lo dice, igual que
YouTube sin llave. El tablero tiene que poder mostrar prensa sin redes.
"""

import json
import os
from urllib.error import HTTPError
from urllib.parse import urlencode
from urllib.request import Request, urlopen

from . import VERSION
from .entorno import cargar, primero

API = "https://api.apify.com/v2/"
AGENTE = "PulsoN33/{}".format(VERSION)

# Tope de la plataforma para run-sync-get-dataset-items. No es configurable.
TOPE_SINCRONO = 300

# Claves de entrada que significan 'esto corre logueado'. Una entrada que
# traiga cualquiera de estas es un error de configuracion, no un aviso.
# NO se listan 'username' ni 'user': en la mayoria de los actores son el
# perfil objetivo y prohibirlas dejaria sin usar todo el catalogo publico.
LLAVES_DE_SESION = frozenset({
    "accesstoken", "authtoken", "bearertoken", "c_user", "cookie", "cookies",
    "csrftoken", "li_at", "login", "logincredentials", "logins", "password",
    "sessionid", "sessionids", "sessioncookie", "sessioncookies", "sessionpool",
    "sessiontoken", "xf_session",
})


class SinToken(Exception):
    """Falta APIFY_TOKEN. Se degrada a vacio, no se revienta la corrida."""


class ActorProhibido(Exception):
    """La entrada del actor pide sesion. Ver la regla 1 del encabezado."""


class PresupuestoAgotado(Exception):
    """Se acabaron los resultados pagados de esta corrida."""


def token(entorno=None):
    """Lee el token del entorno o de un .env. Nunca se imprime ni va a data/.

    Acepta los dos nombres que circulan: Apify documenta `APIFY_TOKEN` pero su
    consola lo ofrece como `APIFY_API_TOKEN`. Ver pulso/entorno.py.
    """
    # Un `entorno` explicito ES el entorno completo y no se le suma ningun
    # .env: si no, una prueba que pasa {} para probar la degradacion acabaria
    # leyendo el token real del disco y no probaria nada.
    t = primero("APIFY_TOKEN", "APIFY_API_TOKEN",
                entorno=cargar() if entorno is None else dict(entorno))
    if not t:
        raise SinToken(
            "falta APIFY_TOKEN (o APIFY_API_TOKEN). Se saca en "
            "console.apify.com > Settings > API & Integrations, y se lee del "
            "entorno o de un .env. Es un token personal: no lo pegues en el "
            "repo ni en un archivo de configuracion.")
    return t


def revisar_entrada(entrada, actor=""):
    """Rechaza una entrada que traiga credenciales o cookies de sesion.

    Compara en minusculas porque los actores no se ponen de acuerdo entre
    `sessionCookie`, `sessioncookie` y `SessionCookies`.
    """
    encontradas = sorted(
        k for k in (entrada or {}) if str(k).strip().lower() in LLAVES_DE_SESION)
    if encontradas:
        raise ActorProhibido(
            "el actor {} recibe {}, o sea que correria con sesion iniciada. "
            "docs/PLAN.md seccion 3 lo refusa: iniciar sesion convierte al "
            "operador en 'usuario' y lo somete a los terminos de la "
            "plataforma, que es justo la defensa que en Meta v. Bright Data "
            "dependio de NO estar logueado.".format(
                actor or "?", ", ".join(encontradas)))


class Presupuesto:
    """Tope duro de resultados por corrida, repartido entre actores activos.

    Apify cobra por resultado. Sin esto, un actor mal configurado no produce
    una corrida lenta sino un cargo, y el cron corre cuatro veces al dia.
    """

    def __init__(self, resultados=300):
        self.resultados = int(resultados)
        self.gastado = 0
        self.por_concepto = {}

    def reparto(self, cuantos_actores):
        """Cuantos resultados le tocan a cada actor. Reparte, no rifa."""
        if cuantos_actores <= 0:
            return 0
        return max(1, self.resultados // cuantos_actores)

    def cobrar(self, concepto, n):
        if self.gastado + n > self.resultados:
            raise PresupuestoAgotado(
                "el presupuesto de {} resultados se agoto en {}".format(
                    self.resultados, concepto))
        self.gastado += n
        self.por_concepto[concepto] = self.por_concepto.get(concepto, 0) + n

    def resumen(self):
        return {
            "resultados": self.resultados,
            "gastado": self.gastado,
            "por_concepto": dict(sorted(self.por_concepto.items())),
        }


def _pedir(metodo, ruta, tok, cuerpo=None, params=None, timeout=30):
    url = API + ruta
    if params:
        url += "?" + urlencode(params)
    datos = json.dumps(cuerpo).encode("utf-8") if cuerpo is not None else None
    req = Request(url, data=datos, method=metodo, headers={
        "Authorization": "Bearer " + tok,
        "User-Agent": AGENTE,
        "Accept": "application/json",
        "Content-Type": "application/json",
    })
    try:
        with urlopen(req, timeout=timeout) as r:
            crudo = r.read().decode("utf-8")
        return json.loads(crudo) if crudo.strip() else []
    except HTTPError as e:
        detalle = ""
        try:
            detalle = e.read().decode("utf-8", "replace")[:400]
        except Exception:
            pass
        # 401 es token invalido o revocado; 402 es credito agotado; 408 es el
        # tope sincrono de 300 s. Los tres se distinguen porque piden acciones
        # distintas y el mensaje generico manda a depurar lo que no es.
        if e.code == 401:
            raise SinToken("Apify rechazo el token (401). Revisa que siga "
                           "vigente en console.apify.com.")
        if e.code == 402:
            raise PresupuestoAgotado(
                "Apify reporta credito agotado (402): {}".format(detalle[:200]))
        if e.code == 408:
            raise RuntimeError(
                "el actor paso los {} s del endpoint sincrono. Baja su cuota "
                "en config/apify.json en vez de subir el timeout.".format(
                    TOPE_SINCRONO))
        raise RuntimeError("{} HTTP {}: {}".format(ruta, e.code, detalle[:200]))


def verificar(tok=None, timeout=20):
    """Confirma que el token sirve y devuelve el plan. Nunca devuelve el token.

    Se usa desde `python -m pulso apify --verificar`, que es lo primero que
    hay que correr despues de exportar la variable: un token mal pegado falla
    con 401 en medio del cron, seis horas despues y sin nadie mirando.
    """
    d = _pedir("GET", "users/me", tok or token(), timeout=timeout).get("data", {})
    plan = d.get("plan") or {}
    return {
        "usuario": d.get("username", ""),
        "plan": plan.get("id") or plan.get("name") or "desconocido",
        "de_paga": bool(d.get("isPaying")),
        # El credito mensual es lo unico que dice si el cron va a poder correr
        # los 30 dias del mes. Viene en dolares en el plan.
        "credito_mensual_usd": plan.get("monthlyUsageCreditsUsd"),
    }


def correr_actor(actor, entrada, tok, limite, timeout=None):
    """Corre un actor y devuelve los items del dataset.

    `limite` va en el query y ADEMAS en la entrada cuando el actor la
    respeta: `limit` recorta lo que se descarga, pero el cargo ya se hizo por
    lo que el actor produjo. Cortar solo del lado del cliente es pagar de mas.
    """
    revisar_entrada(entrada, actor)
    segundos = min(int(timeout or TOPE_SINCRONO), TOPE_SINCRONO)
    ruta = "actors/{}/run-sync-get-dataset-items".format(actor.replace("/", "~"))
    items = _pedir("POST", ruta, tok, cuerpo=entrada,
                   params={"timeout": segundos, "limit": limite, "format": "json"},
                   timeout=segundos + 15)
    return items if isinstance(items, list) else []


def leer_catalogo(ruta):
    """Lee config/apify.json y devuelve (activos, errores).

    Un actor apagado se queda en el archivo con su `razon` escrita, como
    sanquintin en config/medios.json: es registro deliberado de un hueco.
    """
    with open(ruta, encoding="utf-8") as f:
        doc = json.load(f)

    errores = []
    activos = []
    for a in doc.get("actores", []):
        nombre = a.get("id", "?")
        if not a.get("activo"):
            continue
        if not a.get("razon"):
            errores.append("{}: activo sin 'razon' escrita".format(nombre))
        if a.get("idioma") not in ("es", "en"):
            errores.append(
                "{}: idioma '{}' no es 'es' ni 'en'. El modelo de tono es "
                "espanol y a texto en otro idioma no devuelve error, devuelve "
                "una etiqueta plausible.".format(nombre, a.get("idioma")))
        try:
            revisar_entrada(a.get("entrada") or {}, nombre)
        except ActorProhibido as e:
            errores.append(str(e))
        activos.append(a)

    return activos, errores

File: web/src/lib/redes-en-vivo/apify.ts
/**
 * Cliente minimo de la API de Apify para la busqueda pagada en vivo. Solo
 * servidor.
 *
 * Es el primer codigo del sitio que llama a Apify. pulso/apify.py usa el
 * extremo SINCRONO (`run-sync-get-dataset-items`), que aqui no sirve: una
 * funcion de Vercel no puede esperar los hasta 300 s de un actor sin
 * arriesgarse a morir a la mitad con la corrida ya cobrada. Aqui se ARRANCA la
 * corrida y se vuelve; el navegador pregunta cada pocos segundos y cada
 * pregunta mira el estado (responder.ts).
 *
 * Dos cosas que el cliente de Python no tenia y esta ruta si necesita:
 *
 *  - `maxTotalChargeUsd` en CADA corrida: el tope en dolares lo aplica Apify,
 *    no nosotros despues de pagar. El `Presupuesto` de Python cuenta
 *    resultados y cobra cuando la llamada ya volvio; con una corrida por
 *    boton, un actor que se desboca no puede pasar de su tope.
 *  - Una lista CERRADA de actores (`ACTORES`), cada uno con su razon escrita.
 *    La ruta no acepta un id del navegador: el navegador manda un termino.
 *
 * Y una que si comparte, porque es la frontera legal: `revisarEntrada` es el
 * espejo de pulso/apify.py::revisar_entrada. Ninguna entrada con cookies,
 * credenciales o token de sesion sale de aqui. La busqueda por palabra de
 * Facebook PASA esa revision —su entrada no trae nada de eso— y aun asi es la
 * excepcion que el cliente decidio el 23 de septiembre de 2026: el proveedor
 * busca con cuentas propias. Esta escrita en su fila y en docs/PLAN.md.
 */

export const API_APIFY = "https://api.apify.com/v2";

export interface Actor {
  id: string;
  razon: string;
}

/** Los unicos actores que esta ruta puede llamar. */
export const ACTORES = {
  tiktokBusqueda: {
    id: "clockworks~tiktok-scraper",
    razon: "La busqueda por palabra de TikTok, sin sesion: la misma de config/tiktok.json y de las consultas.",
  },
  tiktokComentarios: {
    id: "clockworks~tiktok-comments-scraper",
    razon: "Los comentarios de los videos que nombran el termino. 0.00075 USD cada uno en el nivel Silver (leido el 23 de septiembre de 2026).",
  },
  instagram: {
    id: "apify~instagram-scraper",
    razon: "La etiqueta del termino y los comentarios de sus publicaciones, sin sesion. Instagram no tiene busqueda por palabra sin sesion.",
  },
  facebookBusqueda: {
    id: "scraper_one~facebook-posts-search",
    razon: "Busqueda por palabra en Facebook. Su entrada no trae sesion, pero la pagina de busqueda de Facebook la exige, asi que el proveedor busca con cuentas propias: cruza la frontera de docs/PLAN.md §3 por decision del cliente del 23 de septiembre de 2026, pendiente de opinion legal.",
  },
  facebookComentarios: {
    id: "apify~facebook-comments-scraper",
    razon: "Los comentarios de las publicaciones publicas que devolvio la busqueda, sin sesion. 0.0017 USD cada uno en el nivel Silver.",
  },
} as const satisfies Record<string, Actor>;

/** Espejo de pulso/apify.py::LLAVES_DE_SESION. `username` y `user` no estan a
 *  proposito: en casi todos los actores son el perfil objetivo. */
export const LLAVES_DE_SESION: ReadonlySet<string> = new Set([
  "accesstoken", "authtoken", "bearertoken", "c_user", "cookie", "cookies",
  "csrftoken", "li_at", "login", "logincredentials", "logins", "password",
  "sessionid", "sessionids", "sessioncookie", "sessioncookies", "sessionpool",
  "sessiontoken", "xf_session",
]);

export class ActorProhibido extends Error {
  constructor(mensaje: string) {
    super(mensaje);
    this.name = "ActorProhibido";
  }
}

export function revisarEntrada(entrada: Record<string, unknown>, actor: string): void {
  const encontradas = Object.keys(entrada).filter((k) => LLAVES_DE_SESION.has(k.trim().toLowerCase())).sort();
  if (encontradas.length > 0) {
    throw new ActorProhibido(
      `el actor ${actor} recibe ${encontradas.join(", ")}, o sea que correria con sesion iniciada; `
      + "docs/PLAN.md seccion 3 lo refusa.",
    );
  }
}

export interface Tope {
  /** Lo mas que Apify puede cobrar por esta corrida. */
  usd: number;
  /** Lo mas que se cobra en resultados, para actores por resultado. */
  items: number;
  /** Segundos antes de que Apify la corte. */
  segundos: number;
}

export type EstadoCorrida = "READY" | "RUNNING" | "SUCCEEDED" | "FAILED" | "TIMING-OUT" | "TIMED-OUT" | "ABORTING" | "ABORTED";

export interface Corrida {
  id: string;
  dataset: string;
  estado: EstadoCorrida;
  /** Lo que de verdad se cobra (`usageTotalUsd`), o null si aun no se sabe. */
  usd: number | null;
}

export const TERMINADAS: ReadonlySet<EstadoCorrida> = new Set(["SUCCEEDED", "FAILED", "TIMED-OUT", "ABORTED"]);

export class ErrorApify extends Error {
  constructor(public readonly estado: number, mensaje: string) {
    super(mensaje);
    this.name = "ErrorApify";
  }
}

const MS_LIMITE = 15_000;

function comoCorrida(crudo: unknown): Corrida {
  const d = (crudo as { data?: Record<string, unknown> } | null)?.data ?? {};
  const id = typeof d.id === "string" ? d.id : "";
  const dataset = typeof d.defaultDatasetId === "string" ? d.defaultDatasetId : "";
  if (!/^[A-Za-z0-9]{8,40}$/.test(id) || !/^[A-Za-z0-9]{8,40}$/.test(dataset)) {
    throw new ErrorApify(502, "Apify devolvio una corrida sin id");
  }
  const usd = typeof d.usageTotalUsd === "number" && Number.isFinite(d.usageTotalUsd) ? d.usageTotalUsd : null;
  return { id, dataset, estado: (typeof d.status === "string" ? d.status : "READY") as EstadoCorrida, usd };
}

async function pedir(url: string, init: RequestInit, token: string, solicitar: typeof fetch): Promise<unknown> {
  const r = await solicitar(url, {
    ...init,
    headers: { ...(init.headers ?? {}), Authorization: `Bearer ${token}` },
    cache: "no-store",
    signal: AbortSignal.timeout(MS_LIMITE),
  });
  // Los mensajes no llevan el cuerpo de Apify: puede traer la entrada, y la
  // entrada lleva el termino que alguien busco.
  if (r.status === 401) throw new ErrorApify(401, "token de Apify rechazado");
  if (r.status === 402) throw new ErrorApify(402, "credito de Apify agotado");
  if (!r.ok) {
    // El TIPO del error si, que no lleva la entrada: fue lo que explico el 400
    // de TikTok de la sonda del 23 de septiembre de 2026 (un tope por debajo
    // del minimo del actor). El mensaje no, que puede citar la entrada.
    const tipo = await r.json().then((c: { error?: { type?: unknown } }) => (typeof c?.error?.type === "string" ? c.error.type : null)).catch(() => null);
    throw new ErrorApify(r.status, `Apify respondio ${r.status}${tipo === null ? "" : ` (${tipo.slice(0, 60)})`}`);
  }
  return r.json();
}

/** Arranca una corrida y vuelve en cuanto Apify la acepta. */
export async function iniciarCorrida(
  actor: Actor,
  entrada: Record<string, unknown>,
  tope: Tope,
  token: string,
  solicitar: typeof fetch = fetch,
): Promise<Corrida> {
  revisarEntrada(entrada, actor.id);
  const params = new URLSearchParams({
    timeout: String(tope.segundos),
    maxItems: String(tope.items),
    maxTotalChargeUsd: tope.usd.toFixed(2),
  });
  const cuerpo = await pedir(`${API_APIFY}/acts/${actor.id}/runs?${params.toString()}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(entrada),
  }, token, solicitar);
  return comoCorrida(cuerpo);
}

export async function estadoCorrida(id: string, token: string, solicitar: typeof fetch = fetch): Promise<Corrida> {
  if (!/^[A-Za-z0-9]{8,40}$/.test(id)) throw new ErrorApify(400, "id de corrida invalido");
  return comoCorrida(await pedir(`${API_APIFY}/actor-runs/${id}`, { method: "GET" }, token, solicitar));
}

/** Los items de una corrida. Leerlos no cuesta: el cobro fue al producirlos. */
export async function itemsDe(dataset: string, limite: number, token: string, solicitar: typeof fetch = fetch): Promise<Record<string, unknown>[]> {
  if (!/^[A-Za-z0-9]{8,40}$/.test(dataset)) throw new ErrorApify(400, "id de dataset invalido");
  const params = new URLSearchParams({ clean: "true", format: "json", limit: String(limite) });
  const cuerpo = await pedir(`${API_APIFY}/datasets/${dataset}/items?${params.toString()}`, { method: "GET" }, token, solicitar);
  return Array.isArray(cuerpo) ? cuerpo.filter((x): x is Record<string, unknown> => typeof x === "object" && x !== null) : [];
}
